Feature/#86dqhbza1/show command

LCobham opened the pull request:
# Light-it Pull Request⚡⚡⚡

_Resolve_: [closes #86dqhbza1 ](https://app.clickup.com/t/86dqhbza1)

## Description ⭐

Created a simple console app in the `moviepedia` file, at the root of the project. The `moviepedia` script is quite simple:; it requires the necessary src files, loads the .env file, initializes a symfony console app, registers the commands and runs the app. 

The `src` directory contains the majority of the source code executed when running the `moviepedia` app, namely, the `show` command, which is the only one implemented thus far. The source code for the show command (along with the optional parameters) is located in `showCommand.php`, and this uses the `URL` helper class (in the URL.php file) and the constants (in constants.php). The URL class' main purpose is to provide a simple and elegant way to add more options to the show command in the future. 

Currently, functions.php only contains a simple function that sets the environment variables (which could be improved to skip commented lines or invalid lines).

### Type of change 🏃

- [ ] Bug fix (non-breaking change which fixes an issue)
- [✅] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] This change requires a documentation update

### Anything Else? 🙈

A "known bug" is that, when providing the `fullPlot` option, the format of the table printed usually gets messed up since the text is too large for a single line (but this is a "bug" of the symfony/console Table class). Perhaps the code could be modified so that the description is not displayed inside the table.

LCobham committed:
feat: Create v1.0 of console with show command and options.

LCobham committed:
core: Add simple README file

ezef commented:
> A "known bug" is that, when providing the fullPlot option, the format of the table printed usually gets messed up since the text is too large for a single line (but this is a "bug" of the symfony/console Table class). Perhaps the code could be modified so that the description is not displayed inside the table.

Try this one https://www.php.net/manual/en/function.wordwrap.php

ezef commented on src/ShowComand.php:
```diff
@@ -0,0 +1,78 @@
+                "Oops! Looks like something went wrong ({$status})"
+            );
+
+            exit(EXIT_FAILURE);
```
let's do not stop the script execution inside a method, you can return null if the movie was not found

LCobham replied:
Fixed. Returns null if status code is not 200 (note: the movie API returns a 200 code if movie wasn't found)

ezef commented on src/ShowComand.php:
```diff
@@ -0,0 +1,78 @@
+        $status = $response->getStatusCode();
+
+        if ($status !== 200) {
+            $out->writeln(
+                "Oops! Looks like something went wrong ({$status})"
+            );
```
here it would be better to throw an exception, but for now you can return null and let the caller display message

LCobham replied:
Fixed. Also, changed the return of `getMovieInfo` to return status code for proper error handling in `execute` function.

ezef commented on src/ShowComand.php:
```diff
@@ -0,0 +1,78 @@
+            exit(EXIT_FAILURE);
+        }
+
+        return json_decode($response->getBody());
```
One thing you can do is create a Movie class with Title and info properties ( I think those are the only ones used all around the code), and then use it all around the code

LCobham replied:
Fixed. (See `src/Movie.php` and `displayAsTable` function in `src/showCommand.php`)

LCobham committed:
fix: Fix known bug in fullPlot display

LCobham committed:
refactor: Create Movie class & modify getMovieInfo error handling

ezef commented on src/ShowComand.php:
```diff
@@ -49,27 +54,28 @@ protected function getMovieInfo(string $url, OutputInterface $out)
         }
 
-        return json_decode($response->getBody());
+        return [new Movie($response), $status];
```
I would rather stick to return the Movie or null. Think about isolate things in one method. If you return the Movie or null, you are encapsulating all the logic about API call to fetch movie inside this method and you are not leaking information (returnting the status code) about how it's implemented to the users of this method

ezef commented on src/ShowComand.php:
```diff
@@ -28,11 +28,16 @@ public function execute(InputInterface $input, OutputInterface $output)
-        $movieInfo = $this->getMovieInfo($url, $output);
+        [$movieInfo, $statusCode] = $this->getMovieInfo($url, $output);
+
+        if (! $movieInfo || $movieInfo->Response !== 'True') {
```
lets move this to inside the getMovieInfo() method, again with the idea of encapsulating all the API related things inside one method

> $movieInfo->Response !== 'True'

ezef commented on src/Movie.php:
```diff
@@ -0,0 +1,29 @@
+    public string $Title, $Year, $Response;
+    public array $info = [];
+
+    public function __construct(ResponseInterface $response)
```
you can use a static method for this instead of using the constructor

> public static function FromResponse(ResponseInterface $response) {
...
...
...

And then you call it like

> Movie::FromResponse($response)

ezef commented on src/Movie.php:
```diff
@@ -0,0 +1,29 @@
+        $this->Year = $parsed->Year;
+        $this->Response = $parsed->Response;
+
+        unset($parsed->Title);
+        unset($parsed->Year);
+        unset($parsed->Response);
```
this is not really needed

ezef commented on src/Movie.php:
```diff
@@ -0,0 +1,29 @@
+        unset($parsed->Response);
+
+        foreach($parsed as $property => $value) {
+            if (gettype($value) === 'string' || gettype($value) === 'int') {
```
PHP have methods to do this checks easier

https://www.php.net/manual/es/function.is-string.php
https://www.php.net/manual/es/function.is-int.php
https://www.php.net/manual/es/function.is-array.php

ezef commented on src/Movie.php:
```diff
@@ -0,0 +1,29 @@
+use Psr\Http\Message\ResponseInterface;
+
+class Movie {
+    public string $Title, $Year, $Response;
```
The Movie Class should not have the `$Response` itself here. 

Think about this: What if next week there is a requirement to fetch movies from DB instead from API Call? is the $response related with a Movie?

LCobham committed:
refactor: Improve separation of concerns in getMovieInfo.

Change getMovieInfo to return null if movie wasnt retrieved; change Movie constructor to static method fromResponse; improved style all over
